src/Form/FileAddArchiveForm.php

```diff
@@ -0,0 +1,152 @@
+      '#required' => TRUE,
+      '#description' => 'Files must be less than <strong>' . format_size($validators['file_validate_size'][0]) . '</strong><br> Allowed file types: <strong>' . $validators['file_validate_extensions'][0] . '</strong>',
+      '#upload_validators' => $validators,
+    );
+
+    $form['pattern'] = array(
+      '#type' => 'textfield',
+      '#title' => $this->t('Pattern'),
+      '#description' => $this->t('Only files matching this pattern will be imported. For example, to import all jpg and gif files, the pattern would be <em>*.jpg|*.gif</em>. Use <em>.*</em> to extract all files in the archive.'),
+      '#default_value' => '.*',
+      '#required' => TRUE,
+    );
+
+    $form['actions'] = array('#type' => 'actions');
+    $form['actions']['submit'] = array(
+      '#type' => 'submit',
+      '#value' => t('Submit'),
+    );
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    if ($archive = File::load($form_state->getValue('upload')[0])) {
+      if ($archiver = archiver_get_archiver($archive->getFileUri())) {
+
+        $extract_dir = file_default_scheme() . '://' . pathinfo($archive->getFilename(), PATHINFO_FILENAME);
+        $extract_dir = file_destination($extract_dir, FILE_EXISTS_RENAME);
+        if (!file_prepare_directory($extract_dir, FILE_MODIFY_PERMISSIONS | FILE_CREATE_DIRECTORY)) {
+          throw new \Exception(t('Unable to prepar, e directory %dir for extraction.', array('%dir' => $extract_dir)));
+        }
+        $archiver->extract($extract_dir);
+        $pattern = '/' . $form_state->getValue('pattern') . '/';
+        if ($files = file_scan_directory($extract_dir, $pattern)) {
+          foreach ($files as $file) {
+            $file = File::create([
+              'uri' => $file->uri,
+              'filename' => $file->filename,
+              'status' => FILE_STATUS_PERMANENT,
+            ]);
+            $file->save();
+          }
+        }
+        drupal_set_message($this->t('Extracted %file and added @count new files.', array('%file' => $archive->getFilename(), '@count' => count($files))));
+      }
+      else {
+        throw new \Exception(t('Cannot extract %file, not a valid archive.', array('%file' => $archive->getFileUri())));
+      }
+    }
+    $this->redirect('entity.file.collection')->send();
+  }
+
+  /**
+   * Retrieves the upload validators for a file.
+   *
+   * @param array $options
+   *   (optional) An array of options for file validation.
+   *
+   * @return array
+   *   An array suitable for passing to file_save_upload() or for a managed_file
+   *   or upload element's '#upload_validators' property.
+   */
+  public function getUploadValidators(array $options = array()) {
```

Comment:
This seems to be very similar to the same function on \Drupal\file_entity\Form\FileAddForm. Can we merge them and create a trait?